File: C#/Refactoring/refactoring_exercise_final/test/za/co/entelect/refactoring_final/domain/BankAccountTest.cs
using NUnit.Framework;
using NUnit.Framework.Constraints;
using refactoring_exercise_final.za.co.entelect.refactoring_final.domain;
using refactoring_exercise_final.za.co.entelect.refactoring_final.exception;

namespace refactoring_exercise_final.test.za.co.entelect.refactoring_final.domain
{
    class BankAccountTest
    {
        public const long InitialBalance = 1000L;

        private BankAccount _savingsAccount;
        private BankAccount _moneyMarketAccount;
        private BankAccount _chequeAccount;

        private BankAccount[] _bankAccounts;

        [Test]
        public void TestPositiveBalance()
        {
            CreateTestAccounts(InitialBalance);

            _savingsAccount.CalculateInterest();
            Assert.That(_savingsAccount.BalanceInCents, new EqualConstraint(1050));

            _chequeAccount.CalculateInterest();
            Assert.That(_chequeAccount.BalanceInCents, new EqualConstraint(1040));

            _moneyMarketAccount.CalculateInterest();
            Assert.That(_moneyMarketAccount.BalanceInCents, new EqualConstraint(1100));
        }

        [Test]
        public void TestNegativeBalanceChequeAccount()
        {
            CreateTestAccounts(-500L);

            _chequeAccount.CalculateInterest();
            Assert.That(_chequeAccount.BalanceInCents, new EqualConstraint(-560));
        }

        [Test]
        public void TestNegativeBalanceSavingsAccount()
        {
            CreateTestAccounts(-500L);
            Assert.Throws<BankAccountException>(() => _savingsAccount.CalculateInterest());
        }

        [Test]
        public void TestNegativeBalanceMoneyMarketAccount()
        {
            CreateTestAccounts(-500L);
            Assert.Throws<BankAccountException>(() => _moneyMarketAccount.CalculateInterest());
        }

        [Test]
        public void TestUpdateBalance()
        {
            CreateTestAccounts(InitialBalance);

            foreach (BankAccount account in _bankAccounts)
            {
                account.CalculateBalance(-500);
                Assert.That(account.BalanceInCents, new EqualConstraint(500));

                account.CalculateBalance(1000);
                Assert.That(account.BalanceInCents, new EqualConstraint(1500));
            }
        }

        [Test]
        public void TestUpdateSavingsAccountBalanceFail()
        {
            CreateTestAccounts(InitialBalance);
            Assert.Throws<BankAccountException>(() => _savingsAccount.CalculateBalance(-InitialBalance * 2));
        }

        [Test]
        public void TestUpdateMoneyMarketAccountBalanceFail()
        {
            CreateTestAccounts(InitialBalance);
            Assert.Throws<BankAccountException>(() => _moneyMarketAccount.CalculateBalance(-InitialBalance * 2));
        }

        private void CreateTestAccounts(long initialBalance)
        {
            _savingsAccount = new SavingsAccount(initialBalance);

            _chequeAccount = new ChequeAccount(initialBalance);

            _moneyMarketAccount = new MoneyMarketAccount(initialBalance);

            _bankAccounts = new[] { _savingsAccount, _chequeAccount, _moneyMarketAccount };
        }
    }
}

File: Java/refactoring_exercise_final/src/test/java/za/co/entelect/refactoring_final/domain/BankAccountTest.java
package za.co.entelect.refactoring_final.domain;

import org.junit.Test;
import za.co.entelect.refactoring_final.exception.BankAccountException;

import static junit.framework.Assert.assertEquals;

public class BankAccountTest {

    public static final long INITIAL_BALANCE = 1000L;

    private BankAccount savingsAccount;
    private BankAccount moneyMarketAccount;
    private BankAccount chequeAccount;

    private BankAccount[] bankAccounts;

    @Test
    public void testPositiveBalance() {
        createTestAccounts(INITIAL_BALANCE);

        savingsAccount.calculateInterest();
        assertEquals(1050, savingsAccount.getBalanceInCents());

        chequeAccount.calculateInterest();
        assertEquals(1040, chequeAccount.getBalanceInCents());

        moneyMarketAccount.calculateInterest();
        assertEquals(1100, moneyMarketAccount.getBalanceInCents());
    }

    @Test
    public void testNegativeBalanceChequeAccount() {
        createTestAccounts(-500L);

        chequeAccount.calculateInterest();
        assertEquals(-560, chequeAccount.getBalanceInCents());
    }

    @Test(expected = RuntimeException.class)
    public void testNegativeBalanceSavingsAccount() {
        createTestAccounts(-500L);
        savingsAccount.calculateInterest();
    }

    @Test(expected = RuntimeException.class)
    public void testNegativeBalanceMoneyMarketAccount() {
        createTestAccounts(-500L);
        moneyMarketAccount.calculateInterest();
    }

    @Test
    public void testUpdateBalance(){
        createTestAccounts(INITIAL_BALANCE);

        for (int i = 0; i < bankAccounts.length; i++) {
            bankAccounts[i].calculateBalance(-500);
            assertEquals(500, bankAccounts[i].getBalanceInCents());

            bankAccounts[i].calculateBalance(1000);
            assertEquals(1500, bankAccounts[i].getBalanceInCents());
        }
    }

    @Test(expected = BankAccountException.class)
    public void testUpdateSavingsAccountBalanceFail(){
        createTestAccounts(INITIAL_BALANCE);
        savingsAccount.calculateBalance(-INITIAL_BALANCE * 2);
    }

    @Test(expected = BankAccountException.class)
    public void testUpdateMoneyMarketAccountBalanceFail(){
        createTestAccounts(INITIAL_BALANCE);
        moneyMarketAccount.calculateBalance(-INITIAL_BALANCE * 2);
    }

    private void createTestAccounts(long initialBalance) {
        savingsAccount = new SavingsAccount(initialBalance
        );

        chequeAccount = new ChequeAccount(initialBalance
        );

        moneyMarketAccount = new MoneyMarketAccount(initialBalance
        );

        bankAccounts = new BankAccount[]{savingsAccount, chequeAccount, moneyMarketAccount};
    }
}
